impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -92,85 +90,116 @@ void FenceWaiterVK::Main() {
   using namespace std::literals::chrono_literals;
 
   while (true) {
-    std::unique_lock lock(wait_set_mutex_);
+    bool terminate;
 
-    // If there are no fences to wait on, wait on the condition variable.
-    wait_set_cv_.wait(lock, [&]() { return !wait_set_.empty() || terminate_; });
+    {
+      std::unique_lock lock(wait_set_mutex_);
 
-    // We don't want to check on fence status or collect wait set entries in the
-    // critical section. Copy the array of entries and immediately unlock the
-    // mutex.
-    WaitSet wait_set = wait_set_;
+      // If there are no fences to wait on, wait on the condition variable.
+      wait_set_cv_.wait(lock,
+                        [&]() { return !wait_set_.empty() || terminate_; });
 
-    const auto terminate = terminate_;
+      // We don't want to check on fence status or collect wait set entries in
+      // the critical section. Copy the array of entries and immediately unlock
+      // the mutex.
+      WaitSet wait_set = wait_set_;
 
-    lock.unlock();
+      // Still under the lock, check if the waiter has been terminated.
+      terminate = terminate_;
+    }
 
     if (terminate) {
+      WaitUntilEmpty();
       break;
     }
 
-    // Check if the context had died in the meantime.
-    auto device_holder = device_holder_.lock();
-    if (!device_holder) {
+    if (!Wait()) {
       break;
     }
+  }
+}
 
-    const auto& device = device_holder->GetDevice();
+void FenceWaiterVK::WaitUntilEmpty() {
```

Comment:
Ditto about not having static thread safety analysis because of condition variables. The locking rules should ideally be static checks instead of comments.

Reply:
Ack.